ipfscluster_test.go

```diff
@@ -816,6 +816,275 @@ func TestClustersReplication(t *testing.T) {
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 1
+		c.config.ReplicationFactorMax = nClusters
+	}
+
+	// Shutdown two peers
+	clusters[nClusters-1].Shutdown()
+	clusters[nClusters-2].Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	delay()
+
+	f := func(t *testing.T, c *Cluster) {
+		if c == clusters[nClusters-1] || c == clusters[nClusters-2] {
+			return
+		}
+		p, err := c.PinGet(h)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if len(p.Allocations) != nClusters-2 {
+			t.Error("should have pinned nClusters - 1 allocations")
+		}
+
+		if p.ReplicationFactorMin != 1 {
+			t.Error("rplMin should be 1")
+		}
+
+		if p.ReplicationFactorMax != nClusters {
+			t.Error("rplMax should be nClusters")
+		}
+	}
+	runF(t, clusters, f)
+}
+
+// This test checks that we do not pin something for which
+// we cannot reach ReplicationFactorMin
+func TestClustersReplicationFactorMin(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = nClusters - 1
+		c.config.ReplicationFactorMax = nClusters
+	}
+
+	// Shutdown two peers
+	clusters[nClusters-1].Shutdown()
+	clusters[nClusters-2].Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err == nil {
+		t.Error("Pin should have failed as rplMin cannot be satisfied")
+	}
+	t.Log(err)
+	if !strings.Contains(err.Error(), fmt.Sprintf("not enough peers to allocate CID")) {
+		t.Fatal(err)
+	}
+}
+
+// This tests checks that repinning something that has becomed
+// underpinned actually changes nothing if it's sufficiently pinned
+func TestClustersReplicationMinMaxNoRealloc(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 1
+		c.config.ReplicationFactorMax = nClusters
+	}
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Shutdown two peers
+	clusters[nClusters-1].Shutdown()
+	clusters[nClusters-2].Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	err = clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err := clusters[0].PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(p.Allocations) != nClusters {
```

Comment:
You could always re-pin increasing the replication_min to max, forcing re-allocation on healthy nodes, and then re-pin reducing the replication_min again. Hmm, actually that might not work because it skips the pin and would not update the replication factor entries. I'll fix that..

Did I understand what you meant though?

Reply:
I believe so yes